### PR TITLE
Remove some `ref` patterns from the compiler

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Previous PR: https://github.com/rust-lang/rust/pull/105368

r? @Nilstrieb

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -533,11 +533,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         layout: Option<TyAndLayout<'tcx>>,
     ) -> InterpResult<'tcx, OpTy<'tcx, M::Provenance>> {
         use rustc_middle::mir::Operand::*;
-        let op = match *mir_op {
+        let op = match mir_op {
             // FIXME: do some more logic on `move` to invalidate the old location
-            Copy(place) | Move(place) => self.eval_place_to_op(place, layout)?,
+            &(Copy(place) | Move(place)) => self.eval_place_to_op(place, layout)?,
```
Comment (Nilstrieb) [inline compiler/rustc_const_eval/src/interpret/operand.rs:538]: Not sure if this is an improvement here, maybe
```suggestion
            Copy(place) | Move(place) => self.eval_place_to_op(*place, *layout)?,
```
would be even nicer?

File: compiler/rustc_const_eval/src/transform/promote_consts.rs [hunk 2]
```diff
@@ -849,54 +849,50 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
...
-                )) => {
-                    // Use the underlying local for this (necessarily interior) borrow.
-                    let ty = local_decls[place.local].ty;
-                    let span = statement.source_info.span;
-
-                    let ref_ty = tcx.mk_ref(
-                        tcx.lifetimes.re_erased,
-                        ty::TypeAndMut { ty, mutbl: borrow_kind.to_mutbl_lossy() },
-                    );
+            let StatementKind::Assign(box (_, Rvalue::Ref(region, borrow_kind, place)))= &mut statement.kind else {
```
Comment (Nilstrieb) [inline compiler/rustc_const_eval/src/transform/promote_consts.rs:852]:
```suggestion
            let StatementKind::Assign(box (_, Rvalue::Ref(region, borrow_kind, place))) = &mut statement.kind else {
```

**Review by Noratrieb — COMMENTED** [review]
Thanks, most of those look way better now! One `let_else` fomatting fix (really excited for the day when rustfmt will finally work with it :D) and a nit that you may or may not address, whatever you prefer.
r=me then

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 3]
```diff
@@ -167,7 +167,7 @@ pub(super) fn op_to_const<'tcx>(
         }
     };
     match immediate {
-        Left(ref mplace) => to_const_value(mplace),
+        Left(mplace) => to_const_value(&mplace),
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:170]: Doesn't this change behavior? Previously we took a reference to the existing `immediate` local, now we move `mplace` to a new location and take a reference to that. Why is that better?

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 4]
```diff
@@ -174,8 +166,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 terminator.kind
             ),
 
-            InlineAsm { template, ref operands, options, destination, .. } => {
-                M::eval_inline_asm(self, template, operands, options)?;
+            InlineAsm { template, operands, options, destination, .. } => {
```
Comment (kadiwa4) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:169]: There are now double references here where there weren't ones previously.

I don't really understand the point of these changes, this `match` expression in particular. Having to dereference values of types that are `Copy` seems like adding complexity to me, not removing it. To me, a lot of these changes seem like very slight improvements at best. :)

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/terminator.rs:169]: Yeah I agree, it's better if the pattern can avoid creating `&Copy` references. Unfortunately due to https://github.com/rust-lang/rust/issues/64586 that often requires `ref` keywords, since match ergonimcs does not work well with `&` patterns. Still I think this is better than having `&Copy` references.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs (hunk 3 above)
Comment (WaffleLapkin) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:170]: I thought that LLVM is smart enough to remove unnecessary moves, but it seems [it is not](https://godbolt.org/z/exWTznxos). I'll revert this change 😓

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_const_eval/src/interpret/operand.rs:538]: I don't think derefing here is nicer. Another alternative is
```suggestion
            &Copy(place) | &Move(place) => self.eval_place_to_op(place, layout)?,
```

**Comment by WaffleLapkin** [comment]
I think I addressed all review comments. @rustbot ready

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 5]
```diff
@@ -45,42 +45,22 @@ pub(super) fn compare_impl_method<'tcx>(
...
-        tcx,
-        impl_m,
-        impl_m_span,
-        trait_m,
-        impl_trait_ref,
-        CheckImpliedWfMode::Check,
-    ) {
-        return;
-    }
+    let _: Result<_, ErrorGuaranteed> = try {
```
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:48]: I appreciate all the small cleanups here, but could you make them in a separate PR next time?

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 6]
```diff
@@ -1242,41 +1211,41 @@ fn compare_number_of_method_arguments<'tcx>(
...
-                let pos = if impl_number_args > 0 { impl_number_args - 1 } else { 0 };
-                if let Some(arg) = impl_m_sig.decl.inputs.get(pos) {
+                })
+            })
+            .or(trait_item_span);
+
+        let impl_span = {
+            let ImplItemKind::Fn(impl_m_sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
+            let pos = impl_number_args.saturating_sub(1);
+            impl_m_sig
```
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:1235]:
```suggestion
        let ImplItemKind::Fn(impl_m_sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
        let pos = impl_number_args.saturating_sub(1);
        let impl_span = impl_m_sig
```
I don't see a point reducing the scope of the bindings here

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 7]
```diff
@@ -961,25 +938,18 @@ fn extract_spans_for_error_reporting<'tcx>(
     trait_m: &ty::AssocItem,
 ) -> (Span, Option<Span>) {
     let tcx = infcx.tcx;
-    let mut impl_args = match tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind {
-        ImplItemKind::Fn(ref sig, _) => {
-            sig.decl.inputs.iter().map(|t| t.span).chain(iter::once(sig.decl.output.span()))
-        }
-        _ => bug!("{:?} is not a method", impl_m),
+    let mut impl_args = {
+        let ImplItemKind::Fn(sig, _) = &tcx.hir().expect_impl_item(impl_m.def_id.expect_local()).kind else { bug!("{:?} is not a method", impl_m) };
```
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:942]: All of these bugs make me think that we should have `expect_*` methods on `ImplItemKind` - but that's something for the future

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs [hunk 8]
```diff
@@ -344,14 +344,14 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
...
-                    Ok(res.try_super_fold_with(self)?)
+                    res.try_super_fold_with(self)?
                 } else {
-                    Ok(res)
+                    res
                 }
             }
 
-            _ => ty.try_super_fold_with(self),
-        })()?;
```
Comment (Nilstrieb) [inline compiler/rustc_trait_selection/src/traits/query/normalize.rs:354]: Very weird, I hope there wasn't something weird going on, but the change looks correct🤨. Well, there are quite a few of those around, I guess it's just an artifact of refactorings and made sense at some point.

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs [hunk 9]
```diff
@@ -230,7 +230,7 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
                         if concrete_ty == ty {
                             bug!(
                                 "infinite recursion generic_ty: {:#?}, substs: {:#?}, \
-                                 concrete_ty: {:#?}, ty: {:#?}",
+                             concrete_ty: {:#?}, ty: {:#?}",
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/query/normalize.rs:233]: whats up w this indentation

**Review by Noratrieb — COMMENTED** [review]
A few more comments, but looking good overall

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs (hunk 5 above)
Comment (WaffleLapkin) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:48]: sure, sorry 😅

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs (hunk 6 above)
Comment (WaffleLapkin) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:1235]: I did this to not pollute the outer scope (e.g. `pos` is not very descriptive), but if you think that's not an issue I can remove the scope

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs (hunk 8 above)
Comment (WaffleLapkin) [reply compiler/rustc_trait_selection/src/traits/query/normalize.rs:354]: Yeah, when I was removing these I was triple checking in case I've missed something ><

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 10]
```diff
@@ -363,11 +363,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         src: &OpTy<'tcx, M::Provenance>,
     ) -> InterpResult<'tcx, Either<MPlaceTy<'tcx, M::Provenance>, ImmTy<'tcx, M::Provenance>>> {
         Ok(match src.as_mplace_or_imm() {
-            Left(ref mplace) => {
-                if let Some(val) = self.read_immediate_from_mplace_raw(mplace)? {
+            Left(mplace) => {
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_const_eval/src/interpret/operand.rs:366]: Thanks

**Comment by WaffleLapkin** [comment]
Here I am hoping that the regression is magically fixed.
@bors try @rust-timer queue

**Comment by WaffleLapkin** [comment]
I have no idea what causes slight perf regression here, should I split it into smaller PRs so that they can be benched separately?...

**Comment by Nilstrieb** [comment]
Running the cachegrind command locally shows that the regression for `keccak` is pretty simple:
```
 529,322,559  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
```
I would suggest reverting the changes around the obligation forest and fulfill

**Comment by WaffleLapkin** [comment]
From what I can see the regressions look like noise, `deep-vector` mostly regressed in LLVM and `coercions` change seems to be in parsing, which I didn't change either. Is that ok?

**Comment by Nilstrieb** [comment]
this does look like noise to me, so we're good to go!

Dear bors,

@bors r+

Love,
@Nilstrieb

**Comment by WaffleLapkin** [comment]
@bors retry
looks spurious (is github borked still?)

**Comment by WaffleLapkin** [comment]
@bors try
The CI failure should be fixed, but I can't test this locally.

**Comment by WaffleLapkin** [comment]
The builder that was failing now passes, so I think this should be ready to be merged.

File: src/tools/compiletest/src/read2.rs [hunk 11]
```diff
@@ -110,9 +110,18 @@ impl ProcOutput {
...
-            ProcOutput::Abbreviated { mut head, skipped, tail } => {
+            ProcOutput::Abbreviated { mut head, mut skipped, tail } => {
+                let mut tail = &*tail;
+
+                // Skip over '{' at the start of the tail, so we don't later wrongfully consider this as json.
+                // See <https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Weird.20CI.20failure/near/321797811>
+                while tail.get(0) == Some(&b'{') {
+                    tail = &tail[1..];
+                    skipped += 1;
+                }
```
Comment (Nilstrieb) [inline src/tools/compiletest/src/read2.rs:121]: this is, uh, certainly something. fun that no one has hit this before, you were the lucky one!

**Comment by Nilstrieb** [comment]
That was quite a PR, as I said already it would be nice if it was better split-up next time. But anyways, should be good to go now!
@bors r+
